File: 121_best_time_to_buy_and_sell_stock/Solution.py
class Solution:
    def maxProfit(self, prices: list) -> int:
        cost, profit = float('inf'), 0
        for price in prices:
            cost = min(cost, price)
            profit = max(profit, price - cost)

        return profit


if __name__ == '__main__':
    prices = [7, 1, 5, 3, 6, 4]
    print(Solution().maxProfit(prices))

File: 121_best_time_to_buy_and_sell_stock/Solution.go
package main

import (
    "fmt"
    "math"
)

func maxProfit(prices []int) int {
    cost, profit := math.MaxInt32, 0
    for _, price := range prices {
        cost = min(cost, price)
        profit = max(profit, price - cost)
    }

    return profit
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

func min(a, b int) int {
    if a < b { return a }
    return b
}

func main() {
    prices := []int{7, 1, 5, 3, 6, 4}
    fmt.Println(maxProfit(prices))
}
